[Feat] 프로젝트 세팅

seokzin opened the pull request:
## 📌 이슈
- close #8 

<br/>

## 📝 설명

Next의 SSR 관련 이슈가 많아서 세팅하는 시간이 오래 걸렸어요. 😥
(커밋 body를 상세히 적었으니 추후 참고!)

- **styled-components 설치 및 기본 css 파일 제거**
  - `babel-plugin-styled-components` 라이브러리를 통해 babel 설정했어요.
  - ssr 및 추가 설정을 해줘야 SSR와 CSR 순간의 클래스명이 일치하게 돼요.
 <br/>  

- **path alias 설정**
  - `tsconfig.json` 파일을 통해 간단하게 처리 가능해요.
  - 추후 custom pathes 확장이 가능하니 참고하세요.
 <br/>

- **불필요한 파일들 추가 정리**
  - 기본 제공하는 css 파일들을 정리했어요.
  - 초기 페이지가 깨끗해졌어요.
 <br/>

- **`next.config.js` 내용 확인 후 제거**
  - React strictmode 관련 설정이라서 제거했어요.
 <br/>

- `_document.ts` 파일 생성
  - styled-component가 렌더링 전에 동작하는 로직을 추가했어요.
  - HTML 헤더 정보들 이 곳에서 관리하면 돼요.
 <br/>

- eslint, prettier 설정
  - ESlint는 Next.js에서 기본 제공되어 쉽게 설정했어요.
  - 반면 Prettier는 굉장히 말썽을 부려서 우선 recommended 옵션으로 세팅했어요..
<br/>

## 📷 스크린샷
![image](https://user-images.githubusercontent.com/43740455/154842169-93ee4008-df6a-4d20-9ca8-6968e26268b0.png)
- normalize 및 global 스타일이 잘 들어갔다요!

seokzin committed:
chore: 불필요한 next.config 파일 삭제

- 사용하지 않는 React strict mode와 관련된 설정이라 삭제함

seokzin committed:
feat: styled-component 라이브러리 설치

- yarn add styled-components @types/styled-components

seokzin committed:
feat: styled-components의 ssr 적용을 위한 babel 설정

- yarn add -D babel-plugin-styled-components
- ssr 적용을 통해 첫 화면에서 css가 의도치 않게 적용되는 문제 해결

seokzin committed:
chore: 서버와 클라이언트의 클래스 해시값 동기화를 위한 설정

- className을 props로 전달할 때 에러가 발생
  - 첫 페이지가 ssr로 생성될 때의 클래스명이 이후 csr로 생긴 클래스명과 불일치하기 때문

- 이를 동기화 하기 위해 다음 설정들을 추가
  - fileName: 코드가 포함된 파일명 알려줌
  - displayName: 클래스명에 해당 스타일 정보 추가
  - pure: 사용하지 않는 속성 제거

seokzin committed:
chore: path alias 설정

seokzin committed:
design: 전역 스타일 설정 및 적용

seokzin committed:
fix: 타이틀 관련 속성을 페이지 레벨로 이동

- `_document.tsx`에서 title 속성 관리시 다음 오류 발생
  - eslint(@next/next/no-title-in-document-head)

seokzin committed:
refac: Home에 의존하는 HTML 속성을 분리

- meta를 제외한 주요 HTML 속성은 App에서 관리
- Home 경로인 `index.tsx` 파일엔 렌더링 역할만 남김

seokzin committed:
feat: HTML 헤더 정보 추가 정의 및 정렬

- title 속성 외에는 전부 `_document.tsx` 파일에서 관리

seokzin committed:
feat: 폰트 관련 HTML 헤더 속성 추가

- `rel="preload"` : 브라우저에 폰트가 필요함을 즉시 알려 폰트 로딩시간
  감소 및 렌더링 성능 향상

devpla commented on .babelrc:
```diff
@@ -0,0 +1,14 @@
+      }
+    ]
+  ]
+}
```
next에서 styled를 쓰려면 SSR 등을 적용해야 한다 메모!

devpla commented on pages/_app.tsx:
```diff
@@ -1,8 +1,22 @@
+import Head from "next/head";
+import { ThemeProvider } from "styled-components";
+import { theme } from "@/styles/theme";
+import GlobalStyle from "@/styles/global";
```
글로벌 스타일 정의 좋아요 👍

devpla commented on pages/_document.tsx:
```diff
@@ -0,0 +1,73 @@
+          <meta property="og:description" content="" />
+          <meta property="og:url" content="//" />
+          <meta name="description" content="" />
+          <meta name="keywords" content="" />
```
상냥해..

devpla commented on pages/_document.tsx:
```diff
@@ -0,0 +1,73 @@
+            href="https//spoqa.github.io/spoqa-han-sans/css/SpoqaHanSansNeo.css"
+            rel="stylesheet"
+            type="text/css"
+          />
```
preload 메모메모

devpla commented on styles/global.ts:
```diff
@@ -0,0 +1,13 @@
+
+  * {
+    box-sizing: border-box;
+  }
```
의견 반영 감사해요 👍

devpla commented on tsconfig.json:
```diff
@@ -13,7 +13,11 @@
+    "baseUrl": ".",
+    "paths": {
+      "@/*": ["./*"]
+    }
```
alias 설정 좋아요 🤗👍

devpla reviewed (changes requested):
코드에서 너무 빛이 나서요 눈부시니까 좀 개선해주세요

seokzin committed:
refac: App 컴포넌트를 함수형으로 변경

- next의 props를 제네릭으로 처리하는 방식으로 올바른 타입 선언
- https://github.com/vercel/next.js/issues/7515

seokzin committed:
design: Spoqa 폰트 글로벌 스타일 임시 적용

seokzin committed:
chore: ESlint 권장 규칙 추가

seokzin committed:
refac: 사용하지 않는 기본 파일 제거

seokzin committed:
feat: TypeScript용 ESlint 설치 및 세팅

seokzin committed:
feat: TypeScript용 Prettier 설치 및 추천 세팅 적용

devpla commented on package.json:
```diff
@@ -9,15 +9,22 @@
     "eslint-config-next": "12.1.0",
+    "eslint-config-prettier": "^8.4.0",
+    "eslint-plugin-prettier": "^4.0.0",
+    "prettier": "^2.5.1",
```
prettier랑 eslint가 생겼군요

seokzin replied:
꼼꼼한 리뷰 고맙습니다. 감기 조심하세요.

devpla reviewed (approved):
아직도 눈부신데 벽까지 생겼네요 완벽
